services/src/handlers/model_training.rs

```diff
@@ -0,0 +1,199 @@
+    ReservoirSamplingAggregator, SimpleAggregator,
+};
+use crate::workflows::workflow::Workflow;
+
+#[utoipa::path(
+    tag = "Machine Learning",
+    post,
+    path = "/ml/train",
+    request_body = MLTrainRequest,
+    responses(
+        (
+            status = 200, description = "Model training from workflows", body = TaskResponse,
+            example = json!({"taskId": "7f8a4cfe-76ab-4972-b347-b197e5ef0f3c"})
+        )
+    ),
+    security(
+        ("session_token" = [])
+    )
+)]
+pub async fn ml_model_from_workflow_handler<C: ApplicationContext>(
+    session: C::Session,
+    app_ctx: web::Data<C>,
+    info: web::Json<MLTrainRequest>,
+) -> Result<impl Responder> {
+    let ctx = Arc::new(app_ctx.session_context(session));
+
+    let task_id = schedule_ml_model_from_workflow_task(
+        info.input_workflows.clone(),
+        info.label_workflow.clone(),
+        ctx,
+        info.into_inner(),
+    )
+    .await?;
+
+    Ok(web::Json(TaskResponse::new(task_id)))
+}
+
+/// This method initializes the raster operators and produces a vector of typed raster query
+/// processors.
+pub async fn get_query_processors(
+    operators: Vec<Box<dyn RasterOperator>>,
+    exe_ctx: &dyn ExecutionContext,
+) -> Result<Vec<TypedRasterQueryProcessor>> {
+    let initialized = futures::future::join_all(operators.into_iter().map(|op| {
+        let init = op.initialize(exe_ctx);
+        init
+    }))
+    .await
+    .into_iter()
+    .collect::<Vec<_>>();
+
+    let query_processors = initialized
+        .into_iter()
+        .map(|init_raster_op| {
+            let query_processor = init_raster_op?.query_processor()?;
+            Ok(query_processor)
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(query_processors)
+}
+
+pub fn get_operators_from_workflows(
+    workflows: Vec<Workflow>,
+) -> Result<Vec<Box<dyn RasterOperator>>> {
+    let initialized_raster_operators = workflows
+        .into_iter()
+        .map(|workflow| {
+            let op = workflow.operator;
+            let raster_op = op.get_raster()?;
+            Ok(raster_op)
+        })
+        .collect::<Result<Vec<_>, Error>>()?;
+    Ok(initialized_raster_operators)
+}
+
+/// This method forwards the collected tile data to the appropriate implementation of
+/// the actual aggregation algorithm.
+fn accumulate_tile_data(
+    tile_data: Vec<Option<f32>>,
+    accumulator: &mut Box<dyn Aggregatable<Data = Vec<f32>> + Send>,
+) {
+    let v = tile_data.into_iter().flatten().collect::<Vec<f32>>();
+
+    accumulator.aggregate(v);
+}
+
+//TODO: add a way to abort the query execution when the tasks is aborted
+/// Build ML Features from the raw data and assign feature names.
+pub async fn accumulate_raster_data(
+    feature_names: Vec<Option<String>>,
+    processors: Vec<TypedRasterQueryProcessor>,
+    query: VectorQueryRectangle,
+    ctx: &dyn QueryContext,
+    accum_variant: MachineLearningAggregator,
```

Comment:
instead of passing the aggregator variant, maybe pass the concrete aggregator directly as a generic. So make this method generic over the concrete aggragator type and then just use the supplied aggregator

```
pub async fn accumulate_raster_data<A>(
    ...
    aggregator: A) where A: Aggregatable<Data = Vec<f32>> ... {
        aggregator.aggregate(...)

    }